src/player.cpp

```diff
@@ -11410,109 +11513,134 @@ void player::do_read( item *book )
+            if( max_ex < 2 ) {
+                max_ex = 2;
             }
-            if (activity.type != ACT_NULL) {
-                return;
+            if( max_ex > 10 ) {
+                max_ex = 10;
             }
-        }
+            if( max_ex < min_ex ) {
+                max_ex = min_ex;
+            }
+
+            min_ex *= ( originalSkillLevel + 1 ) * elem.second;
+            min_ex = std::max( min_ex, 1 );
+            max_ex *= ( originalSkillLevel + 1 ) * elem.second;
+            max_ex = std::max( min_ex, max_ex );
 
-        int new_skill_level = skill_level;
-        if (new_skill_level > originalSkillLevel) {
-            add_msg(m_good, _("You increase %s to level %d."),
-                    skill.obj().name().c_str(),
-                    new_skill_level);
+            skill_level.readBook( min_ex, max_ex, reading->level );
 
-            if(new_skill_level % 4 == 0) {
-                //~ %s is skill name. %d is skill level
-                add_memorial_log(pgettext("memorial_male", "Reached skill level %1$d in %2$s."),
-                                   pgettext("memorial_female", "Reached skill level %1$d in %2$s."),
-                                   new_skill_level, skill.obj().name().c_str());
+            if( learner->is_player() ) {
+                add_msg( m_info, _( "You learn a little about %s! (%d%%)" ), skill.obj().name().c_str(),
+                         skill_level.exercise() );
+            } else {
+                add_msg( m_info, _( "%s learns a little about %s!" ), learner->disp_name().c_str(),
+                         skill.obj().name().c_str() );
             }
-            lua_callback("on_skill_increased");
-        }
 
-        if( skill_level == reading->level || !skill_level.can_train() ) {
-            add_msg(m_info, _("You can no longer learn from %s."), book->type_name().c_str());
+            if( skill_level == originalSkillLevel ) {
+                if( activity.index == 1 && learner->is_player() ) {
+                    // continuously read until player gains a new skill level
+                    activity.type = ACT_NULL;
+                    read( activity.position );
+                    if( activity.type != ACT_NULL ) {
+                        return;
+                    }
+                }
+            } else {
+                if( learner->is_player() ) {
+                    add_msg( m_good, _( "You increase %s to level %d." ), skill.obj().name().c_str(),
+                             originalSkillLevel + 1 );
+                    if( skill_level % 4 == 0 ) {
+                        //~ %s is skill name. %d is skill level
+                        add_memorial_log( pgettext( "memorial_male", "Reached skill level %1$d in %2$s." ),
+                                          pgettext( "memorial_female", "Reached skill level %1$d in %2$s." ),
+                                          skill_level, skill.obj().name().c_str() );
```

Comment:
`skill_level` is now a reference, not an `int`. You need to explicitly cast it to an `int` here or else some compilers complain.